rtengine/dcp.cc

```diff
@@ -419,7 +426,7 @@ std::map<std::string, std::string> getAliases(const Glib::ustring& profile_dir)
     buffer[read] = 0;
 
     cJSON_Minify(buffer.get());
-    const std::unique_ptr<cJSON> root(cJSON_Parse(buffer.get()));
+    const std::unique_ptr<cJSON, cJSON_deleter> root(cJSON_Parse(buffer.get()));
```

Comment:
If the `cJSON_deleter` was used in another place it had a right to exist. Here a lambda would fit better.

Reply:
If I use a lambda, it will inflate the std::unique_ptr, whereas the `cJSON_deleter` class is stateless so the pointer is free.  In this particular case it won't make a performance difference, but that's why I did it this way.  I don't have a problem changing it.

Reply:
> If I use a lambda, it will inflate the std::unique_ptr, whereas the cJSON_deleter class is stateless so the pointer is free.

Are you sure?

Quoting from Scott Meyers' "Effective Modern C++", item 18:
> I remarked earlier that, when using the default deleter (i.e., delete), you can reasonably assume that std::unique_ptr objects are the same size as raw pointers. When custom deleters enter the picture, this may no longer be the case. Deleters that are function pointers generally cause the size of a std::unique_ptr to grow from one word to two. For deleters that are function objects, the change in size depends on how much state is stored in the function object. Stateless function objects (e.g., from lambda expressions with no captures) incur no size penalty, and this means that when a custom deleter can be implemented as either a function or a captureless lambda expression, the lambda is preferable.

So let's try:

```c++
#include <iostream>
#include <memory>

namespace
{

    struct cJSON_deleter {
        template<typename T>
        void operator()(T *t) {
            delete const_cast<typename std::remove_const<T>::type*>(t);
        }
    };

    void deleteFunc(int* p)
    {
        delete p;
    }

}

int main()
{
    std::cout << sizeof(void*) << std::endl;

    std::unique_ptr<int> default_deleter(new int(5));
    std::cout << sizeof(default_deleter) << std::endl;

    std::unique_ptr<int, cJSON_deleter> cjson_deleter(new int(5));
    std::cout << sizeof(cjson_deleter) << std::endl;

    const auto lambda =
        [](int* p)
        {
            delete p;
        };

    std::unique_ptr<int, decltype(lambda)> lambda_deleter(new int(5), lambda);
    std::cout << sizeof(lambda_deleter) << std::endl;

    std::unique_ptr<int, decltype(&deleteFunc)> function_deleter(new int(5), deleteFunc);
    std::cout << sizeof(function_deleter) << std::endl;

    return 0;
}
```

Here's the output from my 64b system:

```
8
8
8
8
16
```

> In this particular case it won't make a performance difference, but that's why I did it this way. I don't have a problem changing it.

Okay, I see your motivation. Performance (which doesn't matter here) aside, I honestly prefer the most concise solution which would read something like:

```c++
const std::unique_ptr<cJSON, decltype(&cJSON_Delete)> root(cJSON_Parse(buffer.get()), cJSON_Delete);
```

IMHO everything else is nice but needless.

Reply:
Yep, the stateful deleter is small in this case.  I've changed it!